### PR TITLE
[DONE] Carrion monster growing and reproduction

### PR DESCRIPTION
**Pull request by acidia**
Continues and depends upon #23774, monsters can now evolve based on a set time instead of a half-life.  The new age_grow value is independent of the global evolution rate, so chicks always grow to chickens after 14 days (1 season).  Age_grow also behaves differently, evolution half_life assumes all monsters have existed since the cataclysm, so a new zombie will generally evolve to a brute or something the first evolution check.  Age_grow requires a monster to spend X days at each step, regardless of when it is created.  Player doesn't have to witness each step though.  

Input needed, wasn't sure if I should use season length scaling for age_grow.  On the one hand, you should have if for chicks becoming birds on a farm.  On the other, it gives you too much time to deal with a bad infestation.  Roach nymphs can become a problem in 7 days if unchecked, but 14 days is just way too much time to kill all of them.  I went without till we hear if most people object.

Next up, adding a general function to make egg layers drop an egg after some time so proper infestations can occur for factories in #23748.

Small bugfix, 
`const int current_day = to_days<int>( calendar::time_of_cataclysm - calendar::turn );`
Should be
`const int current_day = to_days<int>( calendar::turn - calendar::time_of_cataclysm);`
I think.  I was doing some testing and noticed that no zombies had evolved by year 5.  Recently changed in #23327.  You might suddenly have lots of evolutions if applied.

### CODE REVIEW

**Comment by pisskop** [comment]
> Roach nymphs can become a problem in 7 days if unchecked, but 14 days is just way too much time to kill all of them. I went without till we hear if most people object.

Since this can cover both 'rotting' and 'growing', it seems logical to create a ratio for season_length.  But it would result in absolutely corrupted values for long or short seasons.  I believe that, locally, time is representative of realtime.  Each second is more or less a second, no matter how long a season is.

However, Season length is about how much you can do in a season.  My instinct is that it would be too complex to try to use a third means of balancing the decreased menace of a mon evolving.  Simply tell the player that chickens will be slower to grow and that roaches will also grow slower.  This will result in easier combat and safer yet slower to start up farming.

Solutions to the issue could be lower stats for chickens or  higher number of roaches spawned.  The game just cannot continue to constantly be in a state of indeterminate timeflow though.  if 14 days is the average, we should be focused on balancing for that.  I myself think that 14-21 is right . . .

**Comment by kevingranade** [comment]
> Input needed, wasn't sure if I should use season length scaling for age_grow.

If it were just farm animals I'd say maybe? Since it's everything including threats for the player definitely no.
The other thing that seems problematic is that if the player doesn't intervene changing season length is going to adjust population levels.

**Comment by acidia** [comment]
This was a pretty simple change so I'm combining it with the reproduction stuff so I don't have three open pull requests dealing with the same files.

Adds the reproduction section to monster .json's.  Everything mirrors the evolution work previously done.
` 
   "reproduction": {
      "baby_egg": "egg_bird", 
        "baby_count": 2, 
        "baby_timer": 7 
    },`
For #23813, I also allowed "baby_egg" item spawn to be replaced by 
`      "baby_monster": "mon_grouse_chick",`
baby_count works for either items or live spawn but you can't spawn eggs and babies.
baby_timer is in days

We'll eventually need a tags list so we can have active breeding seasons.  The general balance idea is that wild birds only breed in very limited batches while chickens have been selectively breed for egg production year round.

Known issues, live birth generates the monster in the parent's square but not in the active map.  The game then attempts to spawn the baby in the parent's square the next check but since the parent probably hasn't moved it will generate it near the parent.  This will allow babies to clip through its parent's cage.  Its inevitable since two monsters can't occupy the same space, the parent may be in a 1x1 cell, or the creature may drop +10 babies at once.

Also, if you drop a chicken in a hut and leave for a year and come back there will be an array of eggs in a pile each batch 7 (testing) days older than the next as intended.  However, live births will track how many babies should spawn and dumps them all newly born around the parent.  The grouse used for testing ends up with 8 new chicks around her when you show up a year later.  

I was thinking that an interesting application of this outside of carrion and egg reproduction was to use the ability of monsters to drop items at set intervals to make decomposing or crawling zombies drop a chuck of tainted meat occasionally.  This would generate tainted carrion spawning and since plague roaches will avoid combat till they are older and more powerful, small pops of zombies occupying lone buildings that the player approaches infrequently would have its "ecosystem" convert from zombie shamblers to plague roaches over time.  It might be worth adding a filth drop to farm animals now, not just because shit is useful for farming but using the carrion spawn system it would be possible to attract predators to your barns and chicken coops based on shit rotting.  A gamey way of having cleanliness reduce the scent of your animals.

**Comment by BorkBorkGoesTheCode** [comment]
Great idea.

How are you going to regulate the catsplosion caused by all those parthenogenic creatures?

**Comment by acidia** [comment]
Not too worried about a friendly monster catsplosion for a few reasons: most games don't last long enough to get in on the tail of exponential growing especially if you add up all of the times.  ~14 days for a wild egg to rot away, another ~14 days for the bird to mature, then for all but the rare chicken you are going to be waiting till spring/summer of year 2 before your first egg batch once season spawn are in.  If I can get games lasting into year 3 for a catsplosion I'd call that a victory in itself.

If we do go the route of having farm monsters drop a manure item that rots in 6 days to a rich soil item you'll have a negative feedback loop for giant chicken houses.  As long as you run through every 5 days with auto pickup on it won't be an issue but each manure would have a 5% of spawning a coyote or some other vermin-ish predator.  A lone coyote spawning in your chicken coop will slaughter everything if left alone.  Something that would threaten cows would be a little harder to come up with given how strong they are.

Forgetting that you left a pile of butchered zombies in a garage for a year and opening the door to find the entire building filled with plague roaches sounds like intended gameplay.  The solution to bad catsplosions is either smashing them or just walking away (this isn't dwarf fortress, you can build anywhere).

**Comment by DracoGriffin** [comment]
If I'm understanding correctly, items have to be dropped in order for the babies to be born, correct?

What about viviparous animals? That is, the animals that give live birth, rather than egg-laying?

**Comment by acidia** [comment]
Easy, you do a 1-to-1 swap the statement
`"baby_egg": "egg_bird"`
for 
` "baby_monster": "mon_grouse_chick"`
Can't lay an egg and have a baby though.

**Comment by kevingranade** [comment]
I think the prey/predator relationship is going to be a solid fix for this,
additionally we could track resource depletion like grass overgrazing.

Below is theory, not an ask for this pr.

Every non-apex natural creature should have a natural predator whose
population grows when its population grows, and triggers a depletion of the
prey population.  This could either be an explicit feedback loop, or
increases in prey populations could simply trigger increases in predator
population.

Every natural creature with population growth like this should have growth
limited by some limited map resource. I.e. cows and deer are going to
compete for grazing area, and once grazing land is depleted, dependent
populations should crash or at least stop growing. This is especially the
case with apex creatures (no predators).

Unnatural creatures like carrion crawlers might not have either of the
above effects, but they should have some other mechanism, like reproducing
based on a resource that is only generated by player action or similar.
They can also do things like cannibalize their own population to grow, i.e.
more advanced forms are the predator population.

The goal of all the mechanisms is to head off population explosion, so the
simplest version that meets that goal should be sufficient.

**Comment by acidia** [comment]
So how do we want to resolve this merging, biosignature support has been done but I don't know if I should keep adding stuff to this PR or make a third?  I was waiting for  #23774 to merge before resolving conflicts, marking this one as "[DONE]" and opening a third.  Planning on starting work on the flags for reproduction tomorrow.

**Comment by FulcrumA** [comment]
Frankly, I'd be good without grass eating locust and similar. Bigger fauna, common scavengers are the main point of this feature and I'd rather not have it grow too big in scale while still remaining in a single ticket as it seems like another feature just built upon this one. At the same time to again underline, I am not that interested in locust in general, though I eagerly await being able to build and populate my own chicken coop as soon as possible.

File: src/monster.cpp [hunk 1]
```diff
@@ -277,15 +288,16 @@ void monster::try_upgrade(bool pin_time) {
         return;
     }
 
-    const int current_day = to_days<int>( calendar::time_of_cataclysm - calendar::turn );
-
+    //const int current_day = to_days<int>( calendar::time_of_cataclysm - calendar::turn );
```
Comment (BevapDin) [inline src/monster.cpp]: Why is there commented out code?

File: src/monster.cpp [hunk 2]
```diff
@@ -325,6 +337,90 @@ void monster::try_upgrade(bool pin_time) {
...
+        if (baby_timer < 0) {
+            return;
+        }
+        baby_timer += current_day;
+    }
+
+    bool season_spawn = false;
+    bool season_match = true;
+    for( auto &elem : type->baby_flags ) {
+        if( ( elem ) == "SUMMER" || ( elem ) == "WINTER" || ( elem ) == "SPRING" ||
```
Comment (BevapDin) [inline src/monster.cpp]: The parenthesis around `elem` are useless.

File: src/monster.cpp (hunk 1 above)
Comment (acidia) [reply src/monster.cpp]: This is what I suspect was preventing evolutions, which is why I reversed it in the next line.  The only reason I didn't delete it was that I figured there was a small chance that the <> check later in the code could have been reversed, giving a similar effect.

Comment (BevapDin) [reply src/monster.cpp]: If it's wrong code, remove it (removed code can still be accessed via git). If it's correct, keep it.

Commenting it out is just confusing for everyone.

> This is what I suspect was preventing evolutions

Sounds about right. The old code will yield negative number of days.

**Comment by acidia** [comment]
Any tips on how to find what line failed the astyle check in monstergenerator.cpp?  Is there an astyle equivalent of the browser based linter?  Downloaded the program from Sourceforge but have had nothing but problems with it.

File: src/monstergenerator.cpp [hunk 3]
```diff
@@ -938,5 +967,44 @@ void MonsterGenerator::check_monster_definitions() const
...
+            if( !item::type_is_defined( mon.baby_egg ) ) {
+                debugmsg( "item_id %s of monster %s is not a valid item id",
+                          mon.baby_egg.c_str(), mon.id.c_str() );
+            }
+        }
+
+        if( mon.biosignatures ) {
+            if( mon.biosig_timer < 1 ) {
+                debugmsg( "Time between biosignature drops (%d) is invalid for %s",
+                         mon.biosig_timer, mon.id.c_str() );
```
Comment (Night-Pryanik) [inline src/monstergenerator.cpp]: There should be an extra space before `mon.biosig_timer...`

File: src/monstergenerator.cpp [hunk 4]
```diff
@@ -938,5 +967,44 @@ void MonsterGenerator::check_monster_definitions() const
                           mon.upgrade_group.c_str(), mon.id.c_str() );
             }
         }
+
+        if( mon.reproduces ) {
+            if( mon.baby_timer < 1 ) {
+                debugmsg( "Time between reproductions (%d) is invalid for %s",
+                         mon.baby_timer, mon.id.c_str() );
```
Comment (Night-Pryanik) [inline src/monstergenerator.cpp]: There should be an extra space before `mon.baby_timer...`

File: src/monstergenerator.cpp [hunk 5]
```diff
@@ -920,8 +948,9 @@ void MonsterGenerator::check_monster_definitions() const
         }
 
         if( mon.upgrades ) {
-            if( mon.half_life < 0 ) {
-                debugmsg( "half_life %d (< 0) of monster %s is invalid", mon.half_life, mon.id.c_str() );
+            if( mon.half_life < 0 && mon.age_grow < 0 ) {
+                debugmsg( "half_life %d and age_grow %d (<0) of monster %s is invalid",
+                         mon.half_life, mon.age_grow, mon.id.c_str() );
```
Comment (Night-Pryanik) [inline src/monstergenerator.cpp]: Here too.

**Comment by ZhilkinSerg** [comment]
> Any tips on how to find what line failed the astyle check in monstergenerator.cpp? Is there an astyle equivalent of the browser based linter? Downloaded the program from Sourceforge but have had nothing but problems with it.

Most easy way is to astyle the whole file or function that you've added/edited.

Are you on Windows? Visual Studio, Code Blocks or Notepad++?

**Comment by acidia** [comment]
I use Code Blocks and Notepad++ for jsons.  I tried following the instruction in the guide
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/CODE_STYLE.md
but git runs like crap in my windows command prompt and my antivirus fights the astyle program from sourceforge like its the antichrist.

**Comment by acidia** [comment]
Disabling evolution should prevent animal growth at the moment but pushing the factor to .01 effectively stops monster evolution while allowing normal animal growth.  The whole growth mechanics could be separated from evolution but it might make sense to wait until we have enough different timers running on a creature that it is practical to have a single array of <action,alarm_time> so that growth/evolution, animal hunger, animal products, biosignatures, reproduction, special attacks, etc. can be handled by a single function.

**Pull request merged**